kb/development/developer-builds.md

```diff
@@ -0,0 +1,727 @@
+```
+
+### Recreate a build directory
+
+#### Reproduce a build directory
+
+Suppose that `source1/build-foo` contains a build tree with settings that you like, and you want to set up a similar build tree in a different directory `source2`. This is common when making backports, or to share how to reproduce an issue.
+
+With CMake: ???
+
+With mbedtls-prepare-build, the first line of `Makefile` in the build directory contains the command line used to generate it.
+
+```
+$ head -n1 ../source1/build-foo/Makefile
+Generated by /home/johndoe/bin/mbedtls-prepare-build -d build-foo --this --that
+$ mbedtls-prepare-build -d build-foo --this --that
+```
+
+To set up another source directory on the same machine, you can leverage the undocumented variable `SOURCE_DIR`.
+
+```
+make -C ../source1/build-foo/Makefile SOURCE_DIR=$PWD dep
+```
+
+#### Refresh an existing build directory after changing dependencies
+
+Suppose that you have a source tree with an existing build tree, and you modify some C source files in a way that affects the build: new source files, removed source files, or changed include directives.
+
+CMake generally picks up updated dependencies automatically. (Are there limitations?) So just rebuild as usual.
+
+With mbedtls-prepare-build, dependencies are analyzed only when you run the `mbedtls-prepare-build` command. If dependencies have changed, run `make dep` (or its alias `make prepare`) in the build directory.
+
+```
+mbedtls-prepare-build -d build-foo
+make -C build-foo
+# ... hackity hack ...
+# E.g. add a new source file
+make -C build-foo dep
+make -C build-foo
+```
+
+If some files have been removed, you should run `make clean` in the build directory before re-running `make` or `make dep`. Otherwise obsolete build products may remain in the build tree, and they may confuse some scripts that enumerate build products with wildcards.
+
+#### Refresh an existing build directory after changing generated files
+
+CMake automatically rebuilds configuration-independent intermediate files as needed.
+
+Limitation: mbedtls-prepare-build does not understand how the configuration-independent intermediate files are built. If they have changed, you need to regenerate them manually in the source tree first. (On first use, mbedtls-prepare-build generates all the configuration-independent intermediate files, but it doesn't detect when this needs to be re-done.)
+
+In Mbed TLS, to update any intermediate file that needs updating:
+
+```
+make generated_files
+```
+
+In TF-PSA-Crypto, or in Mbed TLS, to unconditionally rebuild all intermediate files:
+
+```
+framework/scripts/make_generated_files.py
+```
+
+If build dependencies have changed in the generated files, [run `make dep` in the build tree](#refresh-an-existing-build-directory-after-changing-dependencies).
+
+#### Refresh an existing build directory after changing branches
+
+After changing branches, or rebasing an existing branch, the build scripts may have changed.
+
+With CMake, ???
```

Comment:
Re-running `cmake ..` in the build directory sometimes works, sometimes not. I'm not sure what to suggest other than removing and re-creating the build directory, but that requires [the ability to reproduce a build directory](https://github.com/Mbed-TLS/mbedtls-docs/pull/180/files#r2107644409).